chain/ton/builder.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// NewNativeTransfer creates a new transfer for a native asset
+func (txBuilder TxBuilder) NewNativeTransfer(from xc.Address, to xc.Address, amount xc.AmountBlockchain, input xc.TxInput) (xc.Tx, error) {
+	txInput := input.(*TxInput)
+
+	var stateInit *tlb.StateInit
+	var err error
+	if txInput.AccountStatus != api.Active {
+		if len(txInput.PublicKey) == 0 {
+			return nil, fmt.Errorf("did not set public-key in tx-input for new ton account %s", from)
+		}
+		stateInit, err = wallet.GetStateInit(ed25519.PublicKey(txInput.PublicKey), tonaddress.DefaultWalletVersion, tonaddress.DefaultSubwalletId)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	toAddr, err := tonaddress.ParseAddress(to)
+	if err != nil {
+		return nil, fmt.Errorf("invalid TON destination %s: %v", to, err)
+	}
+	fromAddr, err := tonaddress.ParseAddress(from)
+	if err != nil {
+		return nil, fmt.Errorf("invalid TON address %s: %v", to, err)
+	}
+	tfMsg, err := BuildTransfer(toAddr, tlb.FromNanoTON(amount.Int()), false, txInput.Memo)
+	if err != nil {
+		return nil, err
+	}
+	msgs := []*wallet.Message{}
+
+	msgs = append(msgs, tfMsg)
+
+	logrus.WithFields(logrus.Fields{
+		"messages":   len(msgs),
+		"state-init": stateInit != nil,
+		"chain":      txBuilder.Asset.GetChain().Chain,
+	}).Debug("building tx")
+
+	cellBuilder, err := BuildV3UnsignedMessage(txInput, msgs)
+	if err != nil {
+		return nil, err
+	}
+
+	return tontx.NewTx(fromAddr, cellBuilder, stateInit), nil
+}
+
+// NewTokenTransfer creates a new transfer for a token asset
+func (txBuilder TxBuilder) NewTokenTransfer(from xc.Address, to xc.Address, amount xc.AmountBlockchain, input xc.TxInput) (xc.Tx, error) {
+	txInput := input.(*TxInput)
+
+	var stateInit *tlb.StateInit
+	var err error
+	if txInput.AccountStatus != api.Active {
+		if len(txInput.PublicKey) == 0 {
+			return nil, fmt.Errorf("did not set public-key in tx-input for new ton account %s", from)
+		}
+		stateInit, err = wallet.GetStateInit(ed25519.PublicKey(txInput.PublicKey), tonaddress.DefaultWalletVersion, tonaddress.DefaultSubwalletId)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	toAddr, err := tonaddress.ParseAddress(to)
+	if err != nil {
+		return nil, fmt.Errorf("invalid TON destination %s: %v", to, err)
+	}
+	fromAddr, err := tonaddress.ParseAddress(from)
+	if err != nil {
+		return nil, fmt.Errorf("invalid TON address %s: %v", to, err)
+	}
+	tokenAddr, err := tonaddress.ParseAddress(txInput.TokenWallet)
+	if err != nil {
+		return nil, fmt.Errorf("invalid TON token address %s: %v", txInput.TokenWallet, err)
+	}
+	amountTlb, err := tlb.FromNano((*big.Int)(&amount), int(txBuilder.Asset.GetDecimals()))
+	if err != nil {
+		return nil, err
+	}
+
+	// Spend max 0.2 TON per Jetton transfer.  If we don't have 0.2 TON, we should
+	// lower the max to our balance less max-fees.
+	maxJettonFee := xc.NewAmountBlockchainFromUint64(200000000)
+	remainingTonBal := txInput.TonBalance.Sub(&txInput.EstimatedMaxFee)
```

Comment:
I'm not sure I fully understand the TON's mechanisms here, but I wonder if we should also account for the fees related to the excess amount return (https://docs.ton.org/develop/dapps/asset-processing/jettons#message-2-1). Or maybe it is already baked into the `txInput.EstimatedMaxFee`?